File: rs/moq-audio/src/playout/sync.rs
//! Where a block lives between being decided and being played.
//!
//! Mirrors `sync_buffer.cc`. Every operation here writes whole pitch periods or whole
//! concealment blocks, none of which are the ten milliseconds the sink asks for, so
//! something has to hold the remainder. This is that something: a run of PCM with a
//! read cursor, keeping a little already played audio behind the cursor because
//! [`expand`](super::expand) and [`merge`](super::merge) both need recent history to
//! work from.
//!
//! It is also where the playhead comes from. [`Self::read`] counts every frame
//! committed to it and [`Self::queued`] counts the ones still waiting, so the media
//! position actually being heard is `read - queued`, which is what an A/V clock has
//! to steer video by. Reading the output frame count instead would drift by exactly
//! the amount every time stretch has moved.

use std::time::Duration;

use super::frames;

/// The engine's output block for one stream.
pub(crate) struct Sync {
	channels: usize,
	/// Interleaved PCM: history, then the cursor, then what is still to play.
	samples: Vec<f32>,
	/// Frame index of the next sample to play.
	cursor: usize,
	/// Played frames kept behind the cursor.
	history: usize,
	/// Frames committed since construction.
	read: u64,
	/// Frames played since construction.
	played: u64,
}

impl Sync {
	/// A buffer holding `history` of already played audio for concealment to study.
	pub(crate) fn new(rate: u32, channels: usize, history: Duration) -> Self {
		Self {
			channels: channels.max(1),
			samples: Vec::new(),
			cursor: 0,
			history: frames(rate, history),
			read: 0,
			played: 0,
		}
	}

	/// Commit interleaved PCM to be played after everything already queued.
	pub(crate) fn push(&mut self, pcm: &[f32]) {
		self.samples.extend_from_slice(pcm);
		self.read += (pcm.len() / self.channels) as u64;
		self.trim();
	}

	/// Play `out.len()` samples, or nothing at all if that many are not queued.
	pub(crate) fn pull(&mut self, out: &mut [f32]) -> bool {
		let wanted = out.len() / self.channels;
		if wanted > self.queued() {
			return false;
		}

		let start = self.cursor * self.channels;
		out.copy_from_slice(&self.samples[start..start + out.len()]);
		self.cursor += wanted;
		self.played += wanted as u64;
		self.trim();
		true
	}

	/// Frames committed but not played yet.
	pub(crate) fn queued(&self) -> usize {
		self.samples.len() / self.channels - self.cursor
	}

	/// Frames committed since construction. The media playhead is this less
	/// [`Self::queued`].
	pub(crate) fn read(&self) -> u64 {
		self.read
	}

	/// Frames played since construction.
	pub(crate) fn played(&self) -> u64 {
		self.played
	}

	/// The most recent `count` frames of played audio, for concealment to study.
	/// Shorter than asked for when that much has not been played yet.
	pub(crate) fn history(&self, count: usize) -> &[f32] {
		let count = count.min(self.cursor);
		&self.samples[(self.cursor - count) * self.channels..self.cursor * self.channels]
	}

	/// Drop everything, played and queued, keeping the frame counters.
	pub(crate) fn flush(&mut self) {
		self.samples.clear();
		self.cursor = 0;
	}

	/// Drop the queued audio without touching the history, for a stream that is about
	/// to be rebuilt from a different point.
	pub(crate) fn truncate(&mut self) {
		self.samples.truncate(self.cursor * self.channels);
	}

	/// Forget played audio past what concealment needs.
	fn trim(&mut self) {
		if self.cursor > self.history {
			let drop = self.cursor - self.history;
			self.samples.drain(..drop * self.channels);
			self.cursor -= drop;
		}
	}
}

#[cfg(test)]
mod tests {
	use super::*;

	fn ramp(from: usize, count: usize) -> Vec<f32> {
		(from..from + count).map(|i| i as f32).collect()
	}

	#[test]
	fn plays_what_it_was_given_in_order() {
		let mut sync = Sync::new(48_000, 1, Duration::from_millis(30));
		sync.push(&ramp(0, 480));
		sync.push(&ramp(480, 480));
		assert_eq!(sync.queued(), 960);

		let mut out = vec![0.0; 480];
		assert!(sync.pull(&mut out));
		assert_eq!(out, ramp(0, 480));
		assert!(sync.pull(&mut out));
		assert_eq!(out, ramp(480, 480));
		assert_eq!(sync.queued(), 0);
		assert_eq!(sync.played(), 960);
	}

	#[test]
	fn refuses_a_pull_it_cannot_fill() {
		let mut sync = Sync::new(48_000, 1, Duration::from_millis(30));
		sync.push(&ramp(0, 240));

		let mut out = vec![0.0; 480];
		assert!(!sync.pull(&mut out));
		assert_eq!(sync.queued(), 240, "a refused pull consumes nothing");
		assert!(out.iter().all(|s| *s == 0.0));
	}

	#[test]
	fn the_playhead_is_what_has_been_read_less_what_is_queued() {
		let mut sync = Sync::new(48_000, 1, Duration::from_millis(30));
		sync.push(&ramp(0, 1440));
		assert_eq!(sync.read() - sync.queued() as u64, 0);

		let mut out = vec![0.0; 480];
		assert!(sync.pull(&mut out));
		assert_eq!(sync.read() - sync.queued() as u64, 480);
	}

	#[test]
	fn keeps_the_history_concealment_needs() {
		let mut sync = Sync::new(48_000, 1, Duration::from_millis(30));
		let mut out = vec![0.0; 480];
		for block in 0..10 {
			sync.push(&ramp(block * 480, 480));
			assert!(sync.pull(&mut out));
		}

		let history = sync.history(1440);
		assert_eq!(history.len(), 1440);
		assert_eq!(history, ramp(4800 - 1440, 1440));
		assert!(sync.history(4800).len() <= 1440, "history is bounded");
	}

	#[test]
	fn stereo_frames_stay_paired() {
		let mut sync = Sync::new(48_000, 2, Duration::from_millis(30));
		sync.push(&[1.0, -1.0, 2.0, -2.0]);
		assert_eq!(sync.queued(), 2);

		let mut out = vec![0.0; 4];
		assert!(sync.pull(&mut out));
		assert_eq!(out, vec![1.0, -1.0, 2.0, -2.0]);
	}

	#[test]
	fn truncate_keeps_the_history_and_drops_the_queue() {
		let mut sync = Sync::new(48_000, 1, Duration::from_millis(30));
		sync.push(&ramp(0, 960));
		let mut out = vec![0.0; 480];
		assert!(sync.pull(&mut out));

		sync.truncate();
		assert_eq!(sync.queued(), 0);
		assert_eq!(sync.history(480), ramp(0, 480));
	}
}

File: js/watch/src/audio/playout/sync.ts
/**
 * Where a block lives between being decided and being played.
 *
 * Mirrors `sync_buffer.cc` and the Rust twin in `rs/moq-audio/src/playout/sync.rs`. Every operation
 * here writes whole pitch periods, none of which are the 128 frames the AudioWorklet asks for, so
 * something has to hold the remainder. This is that something: a run of planar PCM with a read
 * cursor, keeping a little already played audio behind the cursor because a preemptive expand
 * splices onto what has already been heard, and concealment will want the same history.
 *
 * Fixed capacity and no allocation after construction: this runs on the audio thread. That is why
 * the copies are loops rather than `set(subarray())`, since a view is an allocation too.
 */
export class Sync {
	readonly channels: number;
	/** Played frames kept behind the cursor, for an operation to splice onto. */
	readonly history: number;

	// Planar PCM: history, then the cursor, then what is still to play.
	#samples: Float32Array[];
	#length = 0;
	#cursor = 0;

	/**
	 * A buffer holding `history` frames of already played audio plus `block` frames of queued
	 * audio, both per channel.
	 */
	constructor(channels: number, history: number, block: number) {
		this.channels = Math.max(1, channels);
		this.history = history;
		this.#samples = [];
		for (let i = 0; i < this.channels; i++) this.#samples.push(new Float32Array(history + block));
	}

	/** Frames committed but not played yet. */
	get queued(): number {
		return this.#length - this.#cursor;
	}

	/** Frames of already played audio available to splice onto. */
	get played(): number {
		return this.#cursor;
	}

	/** Commit `count` frames of `pcm` to be played after everything already queued. */
	push(pcm: Float32Array[], offset: number, count: number): void {
		this.#compact(count);
		if (this.#length + count > this.#samples[0].length) {
			throw new RangeError(`audio playout: a ${count} frame block does not fit the output buffer`);
		}

		const at = this.#length;
		for (let channel = 0; channel < this.channels; channel++) {
			const src = pcm[Math.min(channel, pcm.length - 1)];
			const dst = this.#samples[channel];
			for (let i = 0; i < count; i++) dst[at + i] = src[offset + i];
		}
		this.#length += count;
	}

	/** Play `count` queued frames into `out` at `offset`. The caller checks {@link queued} first. */
	pull(out: Float32Array[], offset: number, count: number): void {
		if (count > this.queued) throw new RangeError("audio playout: pulled more than was queued");

		const from = this.#cursor;
		for (let channel = 0; channel < out.length; channel++) {
			const src = this.#samples[Math.min(channel, this.channels - 1)];
			const dst = out[channel];
			for (let i = 0; i < count; i++) dst[offset + i] = src[from + i];
		}
		this.#cursor += count;
	}

	/**
	 * Copy the most recent `count` frames of played audio into the front of `dst`.
	 *
	 * Returns how many it copied, which is short of `count` until that much has been played. An
	 * operation that splices onto already played audio starts from this and appends the new media.
	 */
	borrow(dst: Float32Array[], count: number): number {
		const take = Math.min(count, this.#cursor);
		const from = this.#cursor - take;
		for (let channel = 0; channel < dst.length; channel++) {
			const src = this.#samples[Math.min(channel, this.channels - 1)];
			const out = dst[channel];
			for (let i = 0; i < take; i++) out[i] = src[from + i];
		}
		return take;
	}

	/** Drop the queued audio, keeping the history a splice works from. */
	truncate(): void {
		this.#length = this.#cursor;
	}

	/** Drop everything, played and queued. */
	flush(): void {
		this.#length = 0;
		this.#cursor = 0;
	}

	/** Slide the window back so `count` more frames fit, forgetting played audio past the history. */
	#compact(count: number): void {
		if (this.#length + count <= this.#samples[0].length) return;

		const drop = Math.min(this.#cursor, Math.max(0, this.#cursor - this.history));
		if (drop === 0) return;

		for (const plane of this.#samples) plane.copyWithin(0, drop, this.#length);
		this.#length -= drop;
		this.#cursor -= drop;
	}
}
